File: 406-queue-reconstruction-by-height/reconstructQueue.cpp
/**
    思路：排好序后，对于某一个人，>=他的高度的人有几个，在数组中就往后移动几位。
*/
#include <iostream>
#include <vector>

using namespace std;

bool compare(pair<int, int>& a, pair<int, int>& b) {
    return a.first < b.first || (a.first == b.first && a.second > b.second);
}
vector<pair<int, int>> reconstructQueue(vector<pair<int, int>>& people) {
    sort(people.begin(), people.end(), [](pair<int, int>& a, pair<int, int>& b) {
        return a.first < b.first || (a.first == b.first && a.second > b.second);
    });

    int n = people.size();
    for (int i = n - 1; i >= 0; i--) {
        people.insert(people.begin() + i + people[i].second + 1, people[i]);
        people.erase(people.begin() + i);
    }
    return people;
}
int main() {
    vector<pair<int, int>> people = {{7,0}, {4,4}, {7,1}, {5,0}, {6,1}, {5,2}};
    reconstructQueue(people);
    return 0;
}

File: 406-queue-reconstruction-by-height/reconstructQueue.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

class Solution(object):
    def reconstructQueue(self, people):
        """
        :type people: List[List[int]]
        :rtype: List[List[int]]
        """
        people.sort(key = lambda row: (row[0], -row[1]))
        n = len(people)
        for i in range(n - 2, -1, -1):
            people.insert(i + people[i][1] + 1, people[i])
            people.remove(people[i])
        return people

s = Solution()
print(s.reconstructQueue([[7,0], [4,4], [7,1], [5,0], [6,1], [5,2]]))
